octodns/provider/azuredns.py

```diff
@@ -0,0 +1,369 @@
+                raise
+
+    def populate(self, zone, target=False, lenient=False):
+        '''
+            Required function of manager.py.
+
+            Special notes for Azure.
+            Azure zone names omit final '.'
+            Azure root records names are represented by '@'. OctoDNS uses ''
+            Azure records created through online interface may have null values
+            (eg, no IP address for A record).
+            Azure online interface allows constructing records with null values
+            which are destroyed by _apply.
+
+            Specific quirks such as these are responsible for any non-obvious
+            parsing in this function and the functions '_params_for_*'.
+
+            :param zone: A dns zone
+            :type  zone: octodns.zone.Zone
+            :param target: Checks if Azure is source or target of config.
+                           Currently only supports as a target. Does not use.
+            :type  target: bool
+            :param lenient: Unused. Check octodns.manager for usage.
+            :type  lenient: bool
+
+            :type return: void
+        '''
+        zone_name = zone.name[0:len(zone.name) - 1]
+        self.log.debug('populate: name=%s', zone_name)
+        before = len(zone.records)
+
+        self._populate_zones()
+        self._check_zone(zone_name)
+
+        _records = set()
+        records = self._dns_client.record_sets.list_by_dns_zone
+        for azrecord in records(self._resource_group, zone_name):
+            if azrecord.type in self.SUPPORTS:
+                _records.add(azrecord)
+        for azrecord in _records:
+            record_name = azrecord.name if azrecord.name != '@' else ''
+            data = getattr(self, '_data_for_{}'.format(azrecord.type))
+            data = data(azrecord)
+            data['type'] = azrecord.type
+            data['ttl'] = azrecord.ttl
+            record = Record.new(zone, record_name, data, source=self)
+            zone.add_record(record)
+
+        self.log.info('populate: found %s records', len(zone.records) - before)
+
+    def _data_for_A(self, azrecord):
+        return {'values': [ar.ipv4_address for ar in azrecord.arecords]}
+
+    def _data_for_AAAA(self, azrecord):
+        return {'values': [ar.ipv6_address for ar in azrecord.aaaa_records]}
+
+    def _data_for_TXT(self, azrecord):
+        return {'values': [reduce((lambda a, b: a + b), ar.value)
+                           for ar in azrecord.txt_records]}
+
+    def _data_for_CNAME(self, azrecord):
+        '''Parsing data from Azure DNS Client record call
+            :param azrecord: a return of a call to list azure records
+            :type  azrecord: azure.mgmt.dns.models.RecordSet
+
+            :type  return: dict
+
+            CNAME and PTR both use the catch block to catch possible empty
+            records. Refer to population comment.
+        '''
+        try:
+            return {'value': _validate_per(azrecord.cname_record.cname)}
```

Comment:
Some of the changes made last week made the process of pulling data down from providers more "lenient." This may be unneeded now. In fact cause problems when users do a `dump --lenient` which is now intended to pull whatever the provider has, valid as far as octoDNS is concerned or not.